src/main/scala/mesosphere/marathon/state/EntityStoreCache.scala

```diff
@@ -0,0 +1,102 @@
+package mesosphere.marathon.state
+
+import mesosphere.marathon.LeadershipCallback
+import mesosphere.util.ThreadPoolContext
+import org.slf4j.LoggerFactory
+
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.Future
+
+/**
+  * The entity store cache caches all current versions from the underlying store.
+  * During election we prefetch al unversioned entries.
+  *
+  * Idea:
+  * - old versions (versioned entries) are never cached
+  * - we always cache the current version
+  * - on elected:
+  *    - clear everything
+  *    - load all current versions
+  *  - on defeated:
+  *    - clear everything
+  */
+class EntityStoreCache[T <: MarathonState[_, T]](store: EntityStore[T]) extends EntityStore[T] with LeadershipCallback {
+
+  private[state] val cache = new TrieMap[String, Option[T]]()
+  private[this] implicit val ec = ThreadPoolContext.context
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  override def fetch(key: String): Future[Option[T]] = {
+    if (isUnversionedEntry(key)) {
+      Future.successful{
+        cache.get(key) match {
+          case Some(t) => t
+          case _       => None
+        }
+      }
+    }
+    else {
+      //if we need to fetch a versioned entry, try if this is the latest version we have in the cache
+      //otherwise let the underlying store fetch that entry.
+      val unversioned = key.substring(0, key.indexOf(':'))
+      cache.get(unversioned) match {
+        case Some(Some(t)) if key == unversioned + ':' + t.version => Future.successful(Some(t))
+        case _ => store.fetch(key)
+      }
+    }
+  }
+
+  override def modify(key: String, onSuccess: (T) => Unit = _ => ())(update: Update): Future[T] = {
```

Comment:
If you look into MarathonStore, you see that every change is locked by LockManager. The onSuccess method is called while that lock still is acquired. That means the change to the cache is also synchronized, while the onSuccess handler is called _after_ the store change.